### PR TITLE
abstract an interface

### PR DESCRIPTION
**Pull request by MaggieNgWu**

### CODE REVIEW

File: hsm/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+if(BUILD_SDF)
+    file(GLOB_RECURSE HEADERS "*.h")
+    file(GLOB_RECURSE SRC_LIST "*.cpp")
+    if(BUILD_STATIC)
```
Comment (bxq2011hust) [inline hsm/CMakeLists.txt:4]: https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html

you only need this var and only call add_library once

File: hsm/sdf/SDFCryptoProvider.cpp [hunk 2]
```diff
@@ -0,0 +1,864 @@
...
+        {
+            throw sessionStatus;
+        }
+        m_pool.push_back(sessionHandle);
+    }
+}
+SessionPool::~SessionPool()
+{
+    auto iter = m_pool.begin();
+    while (iter != m_pool.end())
```
Comment (bxq2011hust) [inline hsm/sdf/SDFCryptoProvider.cpp:40]:
```suggestion
    for(auto s : m_pool)
```

File: hsm/sdf/SDFCryptoProvider.cpp [hunk 3]
```diff
@@ -0,0 +1,864 @@
...
+    SGD_HANDLE session = NULL;
+    if (m_pool.size() == 0)
+    {
+        SGD_HANDLE sessionHandle;
+        SGD_RV sessionStatus = SDF_OpenSession(m_deviceHandle, &sessionHandle);
+        if (sessionStatus != SDR_OK)
+        {
+            throw sessionStatus;
+        }
+        m_pool.push_back(sessionHandle);
```
Comment (bxq2011hust) [inline hsm/sdf/SDFCryptoProvider.cpp:58]:
```suggestion
    {
        std::lock_guard<std::mutex> l(mut);
        m_pool.push_back(sessionHandle);
    }
        
```

File: hsm/sdf/SDFCryptoProvider.cpp [hunk 4]
```diff
@@ -0,0 +1,864 @@
...
+    }
+}
+void* SessionPool::GetSession()
+{
+    std::lock_guard<std::mutex> l(mut);
+    SGD_HANDLE session = NULL;
+    if (m_pool.size() == 0)
+    {
+        SGD_HANDLE sessionHandle;
+        SGD_RV sessionStatus = SDF_OpenSession(m_deviceHandle, &sessionHandle);
```
Comment (bxq2011hust) [inline hsm/sdf/SDFCryptoProvider.cpp:53]:
```suggestion
        SGD_RV sessionStatus = SDF_OpenSession(m_deviceHandle, &session);
```

File: hsm/sdf/SDFCryptoProvider.h [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+     * Return the instance
+     */
+    static SDFCryptoProvider& GetInstance();
+    static SDFCryptoProvider& GetInstance(int sessionPoolSize);
+
+    /**
+     * Generate key
+     * Return error code
+     */
+    unsigned int KeyGen(AlgorithmType algorithm, Key* key);
```
[the remark this answers is not in the document]
Comment (MaggieNgWu) [reply hsm/sdf/SDFCryptoProvider.h:80]: OK，THX.

**Review by cyjseagull — APPROVED** [review]

**Review by bxq2011hust — APPROVED** [review]

**Pull request merged**